### PR TITLE
AP_Scripting: added applet SM_SwitchMission

### PR DESCRIPTION
**Pull request by WillyZehnder**
Script for switching of missions that are saved on SD-card 'on the fly'.
Selection and loading can be done by one of two different methods (multi-position-switch or push-button).
The amount of missions is theoretically unlimited.
Convenient preparation of environment like searching sufficient table_key, finding correct sub-directory for SITL or SD-card and counting available and checking sufficient number of missions.

### CODE REVIEW

**Comment by timtuxworth** [comment]
What an absolutely brilliant idea!

File: libraries/AP_Scripting/applets/SM_SwitchMission.lua [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+CAUTION: Use this script AT YOUR OWN RISK.
+
+-- Willy Zehnder -- 05.12.2022
+
+]]
+-------- SCRIPT 'CONSTANTS' --------
+local SCRIPT_NAME           = 'SM'          -- abbreviation of scriptname for messages on GCS
+local SUB_DIR_APM           = '/APM/scripts'-- path to scripts on SD-card
+local SUB_DIR_SITL          = './scripts'   -- path to scripts when using SITL
+local SUB_DIR_MISSIONS      = 'missions'    -- subdirectory below scripts-directory where the mission-files are stored
```
Comment (timtuxworth) [inline libraries/AP_Scripting/applets/SM_SwitchMission.lua:26]: Perhaps "missions" should be at the APM root, rather than buried under "scripts"?

**Comment by WillyZehnder** [comment]
> What an absolutely brilliant idea!

Thank you Tim :o)

File: libraries/AP_Scripting/applets/SM_SwitchMission.lua [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+local last_sw_change        = uint32_t(0)
+local file_index            = 0
+local wait_finger_off       = false
+local param_table_key       = 0             -- start-value for searching free or still used table-key between 0 and 200
+
+
+local function send_msg(msg_type, msg)
+  -- wrapper for sending messages to the GCS
+    gcs:send_text(msg_type, string.format('%s: %s', SCRIPT_NAME, msg))
+  end
```
Comment (IamPete1) [inline libraries/AP_Scripting/applets/SM_SwitchMission.lua:67]:
```suggestion
  -- wrapper for sending messages to the GCS
  gcs:send_text(msg_type, string.format('%s: %s', SCRIPT_NAME, msg))
end
```

**Comment by IamPete1** [comment]
If it works and someone else can give it a try I'm happy to approve.

**Comment by WillyZehnder** [comment]
> If it works and someone else can give it a try I'm happy to approve.

The version without script-specific parameters is still commented by some users:
https://discuss.ardupilot.org/t/switch-missions-on-the-fly/92106/31

But I will ask for test-comments of the current version.

**Comment by WillyZehnder** [comment]
> If it works and someone else can give it a try I'm happy to approve.

I improved it and tested the lua in SITL - all is working fine.
McKey tested it in Rover with positive feedback.
https://discuss.ardupilot.org/t/switch-missions-on-the-fly/92106/50

**Comment by WillyZehnder** [comment]
@IamPete1 Hi Pete, now all should be done.

**Comment by WillyZehnder** [comment]
After rebase, it seems the lua-bug of receiving a correct handle when opening a not existing file for reading is fixed.
But that leads to an error in my script because of using a nil-handle.
So I had to rework the script and could simplify the check after file:io.open.

**Comment by WillyZehnder** [comment]
Thank you for rebase @IamPete1 
"test scripting" passed
"I can give it a try here too at some point" - that would be great

**Pull request closed without merging**